### ISSUE TITLE
Better documentation for fundamental concepts.

### ISSUE THREAD

**Issue by traverseda** [verse/verse#35]
I've made a new wiki page that you might want to use to slowly introduce people to verse, and verse concepts.

https://github.com/verse/verse/wiki/Overview

I tried to explain what's up with custom types already, and hopefully it's not too far from the truth.

What confuses me is the difference between tag groups and nodes. That might be a good place to start.  A tag can only belong to one node, but multiple tag groups? Presuming I'm understanding correctly that's not a feature I'd bring to the forefront. I imagine that most of the time someone will want to subscribe to an entire node or just objects in the node, grouping them like that could be useful in some cases, but confuses the issue a fair bit.

It looks to me like you can't attach a tag directly to a node, it needs to go through a tag group? Is that right? I thought you could attach tags to nodes directly.... Looking at the tag command specification it looks like it accepts a node ID.

What use case do tag groups have? Or have I completly misunderstood how they work?

Once I've got that sorted out I'm going to work on explaining tags and nodes in a bit more depth.

**Comment by jirihnidek** [comment]
Hi,
Thanks for creating new Overview page. I will go through the whole text later, but first paragraph looks OK.

There is some explanation of tag groups in specification of tag group commands: https://github.com/verse/verse/wiki/Specification-of-Tag-Group-Commands.

Basically tag groups are intended for sharing structure-like data types. Use case of node and tag group can be following: node can represent 3D object and tag group can represent transformation of this object. Tags in this tag group will be position, rotation and scale. The node can contain another tag group containing some other type of information e.g. meta information, name of object etc. In some cases tag group will contain only one tag and it could looks like useless to have tag groups, but it is wise to have only one way and API to access tags.

**Comment by traverseda** [comment]
Ahh, that makes a lot more sense.

**Comment by traverseda** [comment]
I have a bit more time to dig through this, and I don't mind writing some documentation. Like I said, I'd like to get an example python app running (probably using the minecraft in 500 lines of python project since it's already good for teaching) but before I do that I need to grok the structure a bit better.

So hopefully you don't mind me asking some pretty basic questions.

First off, what's the difference between a tag group and a node? You compare tag groups to python/javascript objects. But in my mind nodes more accurately resemble those structures.

Both the nodes and tag groups seem to be abstractions for the same kind of behavior. Can you explain in what kind of situations you'd want to use one or the other?

**Comment by jirihnidek** [comment]
You are right that nodes are similar to tag groups. Nodes can contain more then tag groups. Nodes can contain layers and other child nodes.

First of all, if you want to share the smallest piece of information (one byte in this case) at Verse server, then you have to use at least one node. This node has to contain tag group and this tag group has to contain tag with some value. You can not use tag without tag group and you can not also use tag group without node.

You can use structure:

    node
     +-> tag_group
            +-> tag1
            +-> tag2
            +-> tag3
            ...
            +-> tagN

to share information about e.g. light.

You can use more complex structure:

    node (object)
     +-> tag_group (transformation)
            +-> tag1 (pos)
            +-> tag2 (scale)
            +-> tag3 (rot)
     +-> child_node (mesh)
          +-> tag_group (material)
                +-> tag1 (diff)
                +-> tag2 (spec)
                +-> tag3 (alpha)
          +-> layer1 (vertices)
          +-> layer2 (triangles)

to share information about 3D object (mesh object)

**Comment by traverseda** [comment]
Right. So a tag_group is, essentially, a leaf node. There's a distinction between container nodes and leaf-nodes for implementations sakes. That makes sense. I imagine you could squeeze out a bit more performance by exposing that to the user, but it does complicate the abstraction a bit.

Layers are also a leaf node? You attach tags to them directly? Or do you first create a tag_group, then attach a tag group to a layer?

I guess the big question is what do you expect to use layers for? I'm presuming there are reasons why the layers require all data to be of the same data type.

**Comment by jirihnidek** [comment]
Hi, in terms of tree graph, any tag is "leaf node" (not Verse node). Layers are "leaf nodes" in similar manner.

You can not attach tags or tag group to layer. Layers can be attached to nodes. Layer could be used as associative array, but you can save only one data type in the layer and you can use only unsigned integers for indexing. You can create "holes", when you set values in layer, e.g.:

    layer.values[7] = 2.3;
    layer.values[13] = -1.45;

Layer are expected for storing large data sets, like position of vertices, indexes of triangles, indexes of edges, pixels of images, etc. All data of the layer have to be same data type, because design of protocol: https://github.com/verse/verse/wiki/Specification-of-Layer-Commands
